terraform/main.tf

```diff
@@ -0,0 +1,41 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+resource "juju_application" "opensearch" {
+
+  charm {
+    name     = "opensearch"
+    channel  = var.channel
+    revision = var.revision
+    base     = var.base
+  }
+  config      = var.config
+  model       = var.model
+  name        = var.app_name
+  units       = var.units
+  constraints = var.constraints
+
+
+  # TODO: uncomment once final fixes have been added for:
+  # Error: juju/terraform-provider-juju#443, juju/terraform-provider-juju#182
+  # placement = join(",", var.machines)
+
+  endpoint_bindings = [
+    for k, v in var.endpoint_bindings : {
+      endpoint = k, space = v
+    }
+  ]
+
+  storage_directives = var.storage
+
+  lifecycle {
+    precondition {
+      condition     = length(var.machines) == 0 || (length(var.machines) > 0 && length(var.machines) == var.units)
```

Comment:
Sorry about that, I also found pretty lame but I was getting a bunch of errors because `var.machines` can be null and just `length(var.machines) == var.units` would break...

Anyways, I ended up being hyper-conservative in these conditions!